### PR TITLE
#50877: add support to postgresql_privs to use "FOR { ROLE | USER } target_role" in "ALTER DEFAULT PRIVILEGES"

### LINKED ISSUE #50877: add support to postgresql_privs to use "FOR { ROLE | USER } target_role" in "ALTER DEFAULT PRIVILEGES"
##### SUMMARY
ATM only a workaround (See https://github.com/ansible/ansible/pull/32987#issuecomment-451026860) exists to use the [[ FOR { ROLE | USER } target_role [, ...] ]](https://www.postgresql.org/docs/current/sql-alterdefaultprivileges.html).
I like to see a new parameter **target_roles** like the roles parameter to configure the **target_role** in the default privilege context.
If you don't set the **target_roles** or either not use the *[FOR ROLE ...]* option, the default privileges will be set for the user which is running the command (this will be the *login_user* or if you use *become_user* in your playbook this user).

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
- postgresql_privs

##### ADDITIONAL INFORMATION

A possible playbook could look like this:
```yaml
- name: set default privs
  become_user: postgres
  postgresql_privs:
    db: library
    objs: TABLES,SEQUENCES
    privs: SELECT
    type: default_privs
    role: group_reader
    role_target: reader
    schema: reader
```

### PR DESCRIPTION
**Pull request by tcraxs**
##### SUMMARY

* add support to postgresql_privs to use "FOR { ROLE | USER } target_role" in "ALTER DEFAULT PRIVILEGES"
* Fixes #50877 

ATM only a workaround (See [#32987)](https://github.com/ansible/ansible/pull/32987#issuecomment-451026860) exists to use the [FOR { ROLE | USER } target_role [, ...]](https://www.postgresql.org/docs/current/sql-alterdefaultprivileges.html).
I implement a new parameter **target_roles** like the roles parameter to configure the target_role in the default privilege context.
If you don't set the target_roles or either not use the [FOR ROLE ...] option, the default privileges will be set for the user who is running the command (this will be the login_user or if you use become_user in your playbook this user).

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- postgresql_privs

##### ADDITIONAL INFORMATION

A possible playbook could look like this:
```paste below
- name: set default privs
  become_user: postgres
  postgresql_privs:
    db: library
    objs: TABLES,SEQUENCES
    privs: SELECT
    type: default_privs
    role: group_reader
    role_target: reader
    schema: reader
```

### CODE REVIEW

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (Andersson007) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: Based on official documentation from [postgresql.org],(https://www.postgresql.org/docs/9.4/sql-alterdefaultprivileges.html,)
I suggest to change the description:
```
Comma-separated list of existing role (user/group) names to set as the default permissions
for database objects subsequently created by them.
```
Or something similar

**Comment by Andersson007** [comment]
1. "role_target" is in ADDITIONAL INFORMATION, "target_roles" is in DOCUMENTATION block of source code.
Please, change it and add several items with target_roles to EXAMPLE block of source doce (for both revoke and grand cases)

2. When I executed this playbook, test1 can see content of new tables created by test_role
```
- name: test
  postgresql_privs:
    db: test
    state: absent
    objs: TABLES,SEQUENCES
    privs: SELECT
    type: default_privs
    role: test1
    target_roles: test_role
    schema: public
```
Please, check all use cases for right work.
3. Ensure that target_roles does make sense with any existing parameter.

**Review by Andersson007 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (tcraxs) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py]: Thanks for the hint, I changed that and add examples to the code, how to use target_roles

**Comment by tcraxs** [comment]
> 1. "role_target" is in ADDITIONAL INFORMATION, "target_roles" is in DOCUMENTATION block of source code.

I'm not sure what do you mean with 
> "role_target" is in ADDITIONAL INFORMATION

>    Please, change it and add several items with target_roles to EXEMPLE block of source doce (for both revoke and grand cases)

Added new EXEMPLE block for grant and revoke cases.

> Please, check all use cases for right work.
> 3. Ensure that target_roles does make sense with any existing parameter.

I added a check, that allows `target_roles` only with `type: default_privs`. As far as I see, `target_roles` will only used with `ALTER DEFAULT PRIVILEGES`

**Comment by Andersson007** [comment]
@tcraxs, "role_target" is in ADDITIONAL INFORMATION
The definition of this Issue contains:
```
A possible playbook could look like this:

- name: set default privs
  become_user: postgres
  postgresql_privs:
    db: library
    objs: TABLES,SEQUENCES
    privs: SELECT
    type: default_privs
    role: group_reader
    role_target: reader <<<
    schema: reader
```
But in the source code parameter looks like target_roles

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (Andersson007) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: Please, related you last changes, also change the parameter description "Comma-separated list of existing role (user/group) names to set as the default permissions for database objects subsequently created by them. I(target_roles) is only available with I(type=default_privs)."

**Comment by Andersson007** [comment]
I tried it:
```
 14   - name: set default privs
 15     postgresql_privs:
 16       db: mydb
 17       state: present
 18       objs: TABLES,SEQUENCES
 19       privs: SELECT
 20       type: default_privs
 21       role: test_priv2
 22       target_roles: test_role
 23       schema: public
```
it works, but then I revoke privs by playbook below:
```
 14   - name: set default privs
 15     postgresql_privs:
 16       db: mydb
 17       state: absent
 18       objs: TABLES,SEQUENCES
 19       privs: SELECT
 20       type: default_privs
 21       role: test_priv2
 22       target_roles: test_role
 23       schema: public
```
It returns that nothing has changed and test_priv2 can continue to read the content of new tables created by test_role after that. Shortly, seems state=absent doesn't work.
If I do "ALTER DEFAULT PRIVILEGES FOR ROLE test_role IN SCHEMA public REVOKE SELECT ON TABLES FROM test_priv2;" manually, it works as expected.

Please, check it. Also you should fix shippable tests errors.

**Review by Andersson007 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (tcraxs) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py]: @Andersson007 thanks for your good hints. I forgot the "state=absent" part for this case. I added it and also added your suggestion for the description.

**Comment by tcraxs** [comment]
ready_for_review

**Comment by tcraxs** [comment]
!needs_rebase

**Review by Andersson007 — APPROVED** [review]
I tried it manually. It works as expected. Good job, thank you.
I hope other contributors will approve it soon

@kostiantyn-nemchenko @andytom @astorath

shipit

**Comment by Andersson007** [comment]
@tcraxs if you want to help to improve Postgres support in the next release, https://github.com/ansible/community/wiki/PostgreSQL there are links to several new modules that are waiting for you ;) You can see it in "New PostgreSQL modules" section.
Also we need ci tests for existing modules. You can see them also there.

Would you mind if I add you in this working group ?

Thank you.

**Comment by andytom** [comment]
lgtm
shipit

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (dagwieers) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py]: Please add parameter types to every parameter.
Also, `version_added` are strings.
And it's better to use multiple items in the description if parts are unrelated.
In this case this would be better as:
```yaml
  target_roles:
    description:
    - A list of existing role (user/group) names to set as the
      default permissions for database objects subsequently created by them.
    - Parameter I(target_roles) is only available with C(type=default_privs).
    type: list
    version_added: '2.8'
```

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (dagwieers) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py:317]: Examples should have a `name:` entry.
That is what we recommend to users, so we should set a good example in the docs.

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (dagwieers) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py:343]: Traceback should not be needed, Ansible adds traceback information on errors in except-blocks automatically.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py:343]: @tcraxs these formatting fixes are all that remain. then merge, I hope. There are some issues accumulated for postgresql_privs that I want to start to solve after the merge of this PR. So, please, don't postpone it ;)

Comment (Andersson007) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py:343]: I can start now but, I'm afraid, there will be many merge conflicts with your PR that you / or I need to solve.

**Comment by Andersson007** [comment]
@tcraxs please, fix conflicts, formatting, and add a changelog fragment https://docs.ansible.com/ansible/devel/community/development_process.html#making-your-pr-merge-worthy

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (tcraxs) [reply lib/ansible/modules/database/postgresql/postgresql_privs.py:343]: Sorry I was not able to get some free time for this. I will resolve the merge conflicts right now.
Regarding the "import traceback", I think that is out of my scope. I'm not sure how to resolve that, because it was already in the code before my change. I think we have to rewrite a lot of the code to remove the traceback-thing and port it to match it with the Ansible default error-handling.

**Comment by tcraxs** [comment]
ready_for_review

**Comment by tcraxs** [comment]
bot_status

**Comment by Andersson007** [comment]
@dagwieers , about the examples without 'name', types, traceback, etc. - I'm going to fix it in a separate PR later. If you don't have any notes, please, merge it.

**Comment by Andersson007** [comment]
@tcraxs thank you for this PR. Hope it will be merged.

We also have new several modules and PR that's waiting for review, you can find links to them here https://github.com/ansible/community/wiki/PostgreSQL.
You can write yourself in the table on the working group page. Welcome!

**Comment by Andersson007** [comment]
@chamini2 , could you please look at this PR and approve it by shipit or make notes for this

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (chamini2) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: I would swap the conditionals: `if self._as_who: ... else: `.

Thinking in the current `else` branch is kind of weird: `not not self._as_who`

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (chamini2) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]:
```suggestion
          as_who = None
          if target_roles:
            as_who = ','.join(pg_quote_identifier(r, 'role') for r in target_roles)
```

This `as_who = target_roles` is confusing, since `target_roles` is None, empty string or something falsy in general.

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (chamini2) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: Same as above: swap conditionals.

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (chamini2) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: Same as above: swap conditionals.

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (chamini2) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py]: Same as above: swap conditionals.

File: lib/ansible/modules/database/postgresql/postgresql_privs.py
Comment (chamini2) [inline lib/ansible/modules/database/postgresql/postgresql_privs.py:965]: I am not sure this check is necessary, maybe add a warning of `target_roles` being ignored instead of a failure?

**Review by chamini2 — CHANGES_REQUESTED** [review]

**Comment by Andersson007** [comment]
@chamini2 and we forgot about CI tests.
@tcraxs , this PR implements support of a new functionality, so you need to add CI tests to a new section in test/integration/targets/postgresql/tasks/postgresql_privs.yml by analogy.
If you have any questions about it or need some help, please, let me know.

**Comment by tcraxs** [comment]
> @chamini2 and we forgot about CI tests.
> @tcraxs , this PR implements support of a new functionality, so you need to add CI tests to a new section in test/integration/targets/postgresql/tasks/postgresql_privs.yml by analogy.
> If you have any questions about it or need some help, please, let me know.

Thanks for the good points on swapping conditionals.
I will try my best to straggle through the CI tests, maybe someone will or must help me, because I'm not familiar with such tests.

**Comment by Andersson007** [comment]
@tcraxs , these are usual playbooks, see the files in test/integration/targets/postgresql/tasks/
Also you'll find the useful information in our pinboard https://github.com/ansible/community/issues/435
particularly there are several links in _dagwieers commented on Feb 1_
Try to figure it out.
if there will be difficulties, please, let me know directly by aaklychkov at mail dot ru

**Review by chamini2 — APPROVED** [review]

**Comment by chamini2** [comment]
shipit

**Comment by chamini2** [comment]
Oh, @Andersson007 is right, please add tests!

**Comment by kostiantyn-nemchenko** [comment]
I wonder if it would be possible to add support of managing global default privileges and apply privileges for the whole database, not schema?
If `schema` is specified, the default privileges are altered for objects later created in that schema, otherwise permissions are applied for all newly created objects regardless of schema by omitting `IN SCHEMA` clause in alter statement. 
IMHO would be useful to have such opportunity, though not sure if this feature should be addressed by this PR.

**Comment by tcraxs** [comment]
> Oh, @Andersson007 is right, please add tests!

I'm on it.

**Comment by Andersson007** [comment]
sure, freebsd uses pgsql user (or something like this) instead of postgres

**Comment by Andersson007** [comment]
```
"msg": "relation \"pg_default_acl\" does not exist\nLINE 2:                    FROM pg_default_acl a\n                                ^\n"
```
CentOS 6
there is postgresql 9.0 - very old. You could check a version and use something instead of pg_default_acl if the version is 9.0

**Comment by Andersson007** [comment]
If it is not related with your patch, so just skip it by
```
# Test default_privs with target_role
- include: test_target_role.yml
postgres_version_resp.stdout is version('9.1', '>=')
```
I'm not sure about ('9.1', '>=') or ('9.1', '<=')

**Comment by Andersson007** [comment]
shipit

**Comment by Andersson007** [comment]
@chamini2 , if you're saticfied with the changes, please, write shipit to a new comment

**Comment by Andersson007** [comment]
Ok, @chamini2 , I saw, thank you

**Comment by Andersson007** [comment]
@tcraxs , thank you for this PR! hope you'll stay an active community member. You can help us by review of 51073, e.g. You also can find some interesting ideas here https://github.com/ansible/community/wiki/PostgreSQL or implement your own. I'll be glad to help you with review.
Also we have a community pinboard to communicate about postgres related topics https://github.com/ansible/community/issues/435

**Comment by gundalow** [comment]
Merged into `devel` for release in Ansible 2.8

@tcraxs Thank you for this PR, your first or many I hope. Welcome to the Ansible Community!
@Andersson007 @chamini2 @dagwieers Thank you for the reviews.

@Andersson007 has kindly offered to raise a follow up PR to:
* Add `type:`
* Add `- name:` to all examples

See https://github.com/ansible/ansible/pull/54027 for an example of what this would look like

**Pull request merged**
